File: lib/recalls/normalize.ts
/**
 * Pure string normalization for recall matching. Used identically at seed time
 * (to pre-normalize match_brand / match_product keys) and at query time (to
 * normalize a scanned product), so both sides compare on the same footing.
 *
 * Lowercase, strip diacritics, punctuation/symbols -> single space, collapse
 * whitespace, trim. null / undefined / empty -> "".
 */
export function normalize(s: string | null | undefined): string {
  if (!s) return "";
  return s
    .normalize("NFKD")
    .replace(/[̀-ͯ]/g, "") // strip combining diacritical marks
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, " ") // punctuation / symbols -> space
    .trim()
    .replace(/\s+/g, " ");
}

/** normalize + split into non-empty word tokens. */
export function tokens(s: string | null | undefined): string[] {
  return normalize(s).split(" ").filter(Boolean);
}

/** Digits-only key for barcode comparison (GTIN/EAN). "" if none. */
export function normalizeBarcode(s: string | null | undefined): string {
  if (!s) return "";
  return s.replace(/\D+/g, "");
}

File: scraper/normalize.py
"""Python port of lib/recalls/normalize.ts. MUST stay byte-identical in output to
the TS version — both sides key on the same normalized strings (seed/scrape time
here; query time in TS). Any divergence silently breaks recall matching.

TS pipeline (mirrored exactly): NFKD -> strip combining marks U+0300-U+036F ->
lowercase -> [^a-z0-9]+ -> single space -> trim -> collapse whitespace.
"""
import re
import unicodedata

_COMBINING = re.compile(r"[̀-ͯ]")
_NON_ALNUM = re.compile(r"[^a-z0-9]+")
_WS = re.compile(r"\s+")


def normalize(s: str | None) -> str:
    if not s:
        return ""
    s = unicodedata.normalize("NFKD", s)
    s = _COMBINING.sub("", s)
    s = s.lower()
    s = _NON_ALNUM.sub(" ", s)
    s = s.strip()
    s = _WS.sub(" ", s)
    return s


def normalize_barcode(s: str | None) -> str:
    if not s:
        return ""
    return re.sub(r"\D+", "", s)
